sweep/fee_bumper.go

```diff
@@ -76,4 +92,436 @@ type BumpResult struct {
+		// mempool acceptance.
+		tx, fee, err := t.createAndCheckTx(
+			req.Inputs, req.ChangeScript, f,
+		)
+
+		switch {
+		case err == nil:
+			// Increase the request counter.
+			//
+			// NOTE: this is the only place where we increase the
+			// counter.
+			requestID := t.requestCounter.Add(1)
+
+			// Register the record.
+			t.records.Store(requestID, &monitorRecord{
+				tx:           tx,
+				set:          req.Inputs,
+				changeScript: req.ChangeScript,
+				feeFunction:  f,
+				fee:          fee,
+			})
+
+			return requestID, nil
+
+		// If the error indicates the fees paid is not enough, we will
+		// ask the fee function to increase the fee rate and retry.
+		//
+		// TODO(yy): add this new error in btcwallet to map RBF related
+		// reject reasons:
+		// case errors.Is(err, lnwallet.ErrInsufficientFee):
+		case errors.Is(err, lnwallet.ErrMempoolFee):
+			// If the fee function tells us that we have used up
+			// the budget, we will return an error indicating this
+			// tx cannot be made. The sweeper should handle this
+			// error and try to cluster these inputs differetly.
+			err = f.IncreaseFeeRate()
+			if err != nil {
+				return 0, err
+			}
+
+		// TODO(yy): suppose there's only one bad input, we can do a
+		// binary search to find out which input is causing this error
+		// by recreating a tx using half of the inputs and check its
+		// mempool acceptance.
+		default:
+			return 0, err
+		}
+	}
+}
+
+// createAndCheckTx creates a tx based on the given inputs, change output
+// script, and the fee rate. In addition, it validates the tx's mempool
+// acceptance before returning a tx that can be published directly, along with
+// its fee.
+func (t *TxPublisher) createAndCheckTx(set InputSet, changeScript []byte,
+	f FeeFunction) (*wire.MsgTx, btcutil.Amount, error) {
+
+	// Create the sweep tx with max fee rate of 0 as the fee function
+	// guarantees the fee rate used here won't exceed the max fee rate.
+	//
+	// TODO(yy): refactor this function to not require a max fee rate.
+	tx, fee, err := createSweepTx(
+		set, nil, changeScript, uint32(t.currentHeight),
+		f.FeeRate(), 0, t.cfg.Signer,
+	)
+	if err != nil {
+		return nil, 0, fmt.Errorf("create sweep tx: %v", err)
+	}
+
+	// Sanity check the budget still covers the fee.
+	if fee > set.Budget() {
+		return nil, 0, fmt.Errorf("%w: budget=%v, fee=%v",
+			ErrNotEnoughBudget, set.Budget(), fee)
+	}
+
+	// TODO(yy): skip this check for neutrino.
+	err = t.TestMempoolAccept(tx, 0)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return tx, fee, nil
+}
+
+// broadcast takes a monitored tx and publishes it to the network. Prior to the
+// broadcast, it will subscribe the tx's confirmation notification and attach
+// the event channel on the record. Any broadcast-related errors will not be
+// returned here, instead, they will be put inside the `BumpResult` and
+// returned to the caller.
+func (t *TxPublisher) broadcast(requestID uint64) (*BumpResult, error) {
```

Comment:
I'm noticing that this publisher doesn't really persist across restarts. Isn't that gonna be a problem? Or am I missing something?

Reply:
The sweeper doesn't really persist any states either. IIUC, the way this'll recover after restart is by using `testmempoolaccept` to determine if a new fee rate step actually needs to happen or not.

Reply:
Yeah maybe we should reload the pending sweep txns on restart and re-offer them to the publisher?

Reply:
I don't see the downside to doing so. Either they get rejected because they are already there and not sufficiently rbf'ed in which case we say "cool thanks 👍🏼" or it accepts it and we say "cool thanks 👍🏼".

Reply:
> Yeah maybe we should reload the pending sweep txns on restart and re-offer them to the publisher?

So today the it's the job of the _resolvers_ to detect that after a restart their inputs weren't swept to re-offer. 


I think the only gap in this would be user issued fee bumping requests.